### ISSUE TITLE
Issue on LS 2.X

### ISSUE THREAD

**Issue by OzWookiee** [HandsetDetection/logstash-filter-handsetdetection#7]
Installed the plugin to my test LS on my local windows machine and I'm getting this error when starting Logstash:

Pipeline aborted due to error {:exception=>#<IO::EAGAINWaitReadable: Unknown error - errno backtraces disabled; run with -Xerrno.backtrace=true to enable>, :backtrace=>[], :level=>:error}

I'm going to turn on backtrace and will post that asap

**Comment by OzWookiee** [comment]
```
[31m
Pipeline aborted due to error 
{
:exception=>#<IO::EAGAINWaitReadable: Unknown error - >, 
:backtrace=>[
    "org/jruby/RubyIO.java:2860:in `read_nonblock'",
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/tcp_timeout-0.1.1/lib/tcp_timeout.rb:96:in `read'",
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/base.rb:443:in `post'",
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/base.rb:353:in `remote'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/hd4.rb:381:in `device_fetch_archive'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/logstash-filter-handsetdetection-4.1.1/lib/logstash/filters/handsetdetection.rb:76:in `register'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/logstash-core-2.3.3-java/lib/logstash/pipeline.rb:182:in `start_workers'", 
     "org/jruby/RubyArray.java:1613:in `each'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/logstash-core-2.3.3-java/lib/logstash/pipeline.rb:182:in `start_workers'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/logstash-core-2.3.3-java/lib/logstash/pipeline.rb:136:in `run'", 
     "C:/Temp/logstash/logstash-2.3.3/vendor/bundle/jruby/1.9/gems/logstash-core-2.3.3-java/lib/logstash/agent.rb:473:in `start_pipeline'"
     ],
:level=>:error
}

[0m
stopping pipeline {:id=>"main"}
```

**Comment by OzWookiee** [comment]
I'm GUESSING here that I need to have the local HandsetDetection installed?

**Comment by elricho** [comment]
Actually - it should work for cloud .. but i wonder if it blocks for too long.

**Comment by OzWookiee** [comment]
Changed to use the online and LS starts fine. Perhaps a more sane error can be thrown?

**Comment by OzWookiee** [comment]
Had to change to LS 2.4.0 and now getting this error:

```
Pipeline main started

[31m Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {"exception"=>#<NoMethodError: undefined method `request' for nil:NilClass>, "backtrace"=>["C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/hd4.rb:331:in `device_detect' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-filter-handsetdetection-4.1.1/lib/logstash/filters/handsetdetection.rb:92:in `filter' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:151:in `multi_filter' ",
 "org/jruby/RubyArray.java:1613:in `each' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:148:in `multi_filter' ",
 "(eval):549:in `initialize' ",
 "org/jruby/RubyArray.java:1613:in `each' ",
 "(eval):543:in `initialize' ",
 "org/jruby/RubyProc.java:281:in `call' ",
 "(eval):567:in `initialize' ",
 "org/jruby/RubyArray.java:1613:in `each' ",
 "(eval):562:in `initialize' ",
 "org/jruby/RubyProc.java:281:in `call' ",
 "(eval):289:in `filter_func' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:267:in `filter_batch' ",
 "org/jruby/RubyArray.java:1613:in `each' ",
 "org/jruby/RubyEnumerable.java:852:in `inject' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:265:in `filter_batch' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:223:in `worker_loop' ",
 "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:201:in `start_workers'"], :level=>:error}
[0m

[31m Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {
    "exception"=>#<NoMethodError: undefined method `request' for nil:NilClass>, 
    "backtrace"=>[
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/hd4.rb:331:in `device_detect' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-filter-handsetdetection-4.1.1/lib/logstash/filters/handsetdetection.rb:92:in `filter' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:151:in `multi_filter' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:148:in `multi_filter' ",
        "(eval):549:in `initialize' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "(eval):543:in `initialize' ",
        "org/jruby/RubyProc.java:281:in `call' ",
        "(eval):567:in `initialize' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "(eval):562:in `initialize' ",
        "org/jruby/RubyProc.java:281:in `call' ",
        "(eval):289:in `filter_func' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:267:in `filter_batch' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "org/jruby/RubyEnumerable.java:852:in `inject' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:265:in `filter_batch' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:223:in `worker_loop' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:201:in `start_workers'"
    ],
    :level=>:error
    }
[0m

[31m Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {
    "exception"=>#<NoMethodError: undefined method `request' for nil:NilClass>,
    "backtrace"=>[
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/hd4.rb:331:in `device_detect' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-filter-handsetdetection-4.1.1/lib/logstash/filters/handsetdetection.rb:92:in `filter' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:151:in `multi_filter' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:148:in `multi_filter' ",
        "(eval):549:in `initialize' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "(eval):543:in `initialize' ",
        "org/jruby/RubyProc.java:281:in `call' ",
        "(eval):567:in `initialize' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "(eval):562:in `initialize' ",
        "org/jruby/RubyProc.java:281:in `call' ",
        "(eval):289:in `filter_func' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:267:in `filter_batch' ",
        "org/jruby/RubyArray.java:1613:in `each' ",
        "org/jruby/RubyEnumerable.java:852:in `inject' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:265:in `filter_batch' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:223:in `worker_loop' ",
        "C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:201:in `start_workers'"
    ], 
    :level=>:error
}
[0m 

NoMethodError: undefined method `request' for nil:NilClass
  device_detect at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/handset_detection-4.1.1/lib/handset_detection/hd4.rb:331
         filter at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-filter-handsetdetection-4.1.1/lib/logstash/filters/handsetdetection.rb:92
   multi_filter at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:151
           each at org/jruby/RubyArray.java:1613
   multi_filter at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/filters/base.rb:148
     initialize at (eval):549
           each at org/jruby/RubyArray.java:1613
     initialize at (eval):543
           call at org/jruby/RubyProc.java:281
     initialize at (eval):567
           each at org/jruby/RubyArray.java:1613
     initialize at (eval):562
           call at org/jruby/RubyProc.java:281
    filter_func at (eval):289
   filter_batch at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:267
           each at org/jruby/RubyArray.java:1613
         inject at org/jruby/RubyEnumerable.java:852
   filter_batch at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:265
    worker_loop at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:223
  start_workers at C:/Temp/logstash/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:201
```

**Comment by OzWookiee** [comment]
Here is the config for the filter: --removed--

**Comment by elricho** [comment]
Thanks @OzWookiee - Better remove those API creds :-) I've grabbed a copy tho :-)

**Comment by elricho** [comment]
@OzWookiee  - Appreciate you giving this plugin a workout too :-)

**Comment by OzWookiee** [comment]
If you've got the config I'll kill the link

**Comment by OzWookiee** [comment]
Same issue in 2.3.3

**Comment by elricho** [comment]
IO::EAGAINWaitReadable issue fixed with tcp_timeout patch.
Undefined method `request' for nil:NilClass fixed in 4.1.3 (or 4.1.2).
